task/control-plane-agent/src/mgs_gimlet/host_phase2.rs

```diff
@@ -0,0 +1,285 @@
+        // in our outgoing net task queue).
+        let port = match current.state {
+            State::NeedToSendFirstMgs(port) => {
+                current.state = State::WaitingForFirstMgs {
+                    port,
+                    deadline: now + DELAY_TRY_OTHER_MGS,
+                };
+                port
+            }
+            State::WaitingForFirstMgs { port, deadline } => {
+                if now < deadline {
+                    return None;
+                }
+                // Timed out waiting for a response from the first MGS we tried;
+                // flip to the other one.
+                let port = match port {
+                    SpPort::One => SpPort::Two,
+                    SpPort::Two => SpPort::One,
+                };
+                current.state = State::WaitingForSecondMgs {
+                    port,
+                    deadline: now + DELAY_RETRY,
+                };
+                port
+            }
+            State::WaitingForSecondMgs { port, deadline } => {
+                if now < deadline {
+                    return None;
+                }
+                // Timed out waiting for a response from the second MGS we
+                // tried; flip back to the first and retry.
+                current.retry_count += 1;
+                if current.retry_count >= MAX_ATTEMPTS {
+                    current.notify_calling_task();
+                    self.current = None;
+                    return None;
+                }
+                let port = match port {
+                    SpPort::One => SpPort::Two,
+                    SpPort::Two => SpPort::One,
+                };
+                current.state = State::WaitingForFirstMgs {
+                    port,
+                    deadline: now + DELAY_TRY_OTHER_MGS,
+                };
+                port
+            }
+            State::Fetched => return None,
+        };
+
+        let message = Message {
+            header: Header {
+                version: gateway_messages::version::V2,
+                message_id,
+            },
+            kind: MessageKind::SpRequest(SpRequest::HostPhase2Data {
+                hash: current.hash,
+                offset: current.offset,
+            }),
+        };
+
+        let n = gateway_messages::serialize(tx_buf, &message).unwrap_lite();
+
+        Some(UdpMetadata {
+            addr: SP_TO_MGS_MULTICAST_ADDR,
+            port: MGS_UDP_PORT,
+            size: n as u32,
+            vid: vlan_id_from_sp_port(port),
+        })
+    }
+
+    pub(crate) fn ingest_incoming_data(
+        &mut self,
+        port: SpPort,
+        hash: [u8; 32],
+        offset: u64,
+        data: &[u8],
+    ) {
+        let current = match self.current.as_mut() {
+            Some(current)
+                if hash == current.hash && offset == current.offset =>
+            {
+                current
+            }
+            _ => return,
```

Comment:
Is this... bad?  Or okay?

Reply:
It's okay, because we're all UDP with retries and resends; it just means MGS sent us data that we don't currently want (presumably it's data we used to want). Imagine something like:

0. Host requests data at offset 0
1. We send a request for offset 0 to MGS 0
2. No response within our timeout window, so we send a request for offset 0 to MGS 1
3. MGS 1 responds with data at offset 0; we forward it back to the host
4. Host requests data at offset N
5. We send a request for offset N to MGS 1
6. MGS 0 finally replies to the request for data at offset 0 we sent back in step 1; we no longer care about this data, so we ignore it here.

Reply:
Added this to a comment in 77e52be88c6ae4563c0dccc98efaea18a2cf9728